Added an optional 'depth' argument to Function.prototype.curry

hughfdjackson opened the pull request:
To close gh-6, I've added the depth parameter described therein to
Function.prototype.curry.

The complete signature now is:

```
curry([depth, [listOfInitialArgs..]])
```

hughfdjackson committed:
Added an optional 'depth' argument to Function.prototype.curry

To close gh-6, I've added the depth parameter described therein to
Function.prototype.curry.

The complete signature now is:

```
curry([depth, [listOfInitialArgs..]])
```

guipn commented:
I like this idea a lot.

The current concern about it is that `f.curry(1, 2, 3);` is misleading. I think we should throw if `arguments.length > 1`.

As discussed and agreed, `f.curry(n)(2, 3);` will be considered clear. Hugh is working on that thought so we can evaluate the implementation.

hughfdjackson committed:
Curry throws an error if passed more than one argument

As per conversations with @guipn, formerly the API allowed calls to be made
like so:

```javascript
    var sum = function(){

        var args = [].slice.call(arguments);

        return args.reduce(function(a, b){ return a + b }, 0);

    };

    var sum5 = sum.curry(5, 2, 3);

```

The first argument would be considered the depth, and the subsequent ones
as arguments (leaving 3 arguments left until the original function is applied.

This was considered confusing, so now the above throws an error.

guipn commented:
For now we think this is ideal. Merged.
